django/core/handlers/base.py

```diff
@@ -228,44 +175,88 @@ def get_response(self, request):
 
         try:
-            # Apply response middleware, regardless of the response
-            for middleware_method in self._response_middleware:
-                response = middleware_method(request, response)
-                # Complain if the response middleware returned None (a common error).
-                if response is None:
-                    raise ValueError(
-                        "%s.process_response didn't return an "
-                        "HttpResponse object. It returned None instead."
-                        % (middleware_method.__self__.__class__.__name__))
+            # This is a noop with new style middlewares!
+            response = self._apply_old_response_middleware(request, response)
             response = self.apply_response_fixes(request, response)
         except Exception:  # Any exception should be gathered and handled
             signals.got_request_exception.send(sender=self.__class__, request=request)
-            response = self.handle_uncaught_exception(request, resolver, sys.exc_info())
+            response = self.handle_uncaught_exception(request, sys.exc_info())
 
         response._closable_objects.append(request)
 
         # If the exception handler returns a TemplateResponse that has not
         # been rendered, force it to be rendered.
-        if not response_is_rendered and callable(getattr(response, 'render', None)):
+        if not getattr(response, 'is_rendered', True) and callable(getattr(response, 'render', None)):
+
             response = response.render()
 
         return response
 
-    def process_exception_by_middleware(self, exception, request):
-        """
-        Pass the exception to the exception middleware. If no middleware
-        return a response for this exception, raise it.
-        """
-        for middleware_method in self._exception_middleware:
-            response = middleware_method(request, exception)
+    def _get_response(self, request):
+        "Returns an HttpResponse object for the given HttpRequest"
+        response = None
+
+        if hasattr(request, 'urlconf'):
+            urlconf = request.urlconf
+            set_urlconf(urlconf)
+            resolver = get_resolver(urlconf)
+        else:
+            resolver = get_resolver()
+
+        resolver_match = resolver.resolve(request.path_info)
+        callback, callback_args, callback_kwargs = resolver_match
+        request.resolver_match = resolver_match
+
+        # Apply view middleware
+        for middleware_method in self._view_middleware:
+            response = middleware_method(request, callback, callback_args, callback_kwargs)
             if response:
                 return response
-        raise
 
-    def handle_uncaught_exception(self, request, resolver, exc_info):
+        wrapped_callback = self.make_view_atomic(callback)
+
+        # TODO: It would be nice if we could move the try/except in _get_response_old,
```

Comment:
Yeah, it'd be nice... the only thing that comes to mind is to set up (probably in `load_middleware`) a `self.inner_exception_handler` or some such attribute, which would be a context manager. If using old-style middleware, `self.inner_exception_handler` would be a context manager that catches exceptions and calls `self.process_exception_by_middleware`. If using new-style middleware, it'd be a no-op context manager.

So you'd still have to wrap this call, and the below render call, in the context manager. I don't know if that's much of an improvement or not - maybe a bit?